python/cuxfilter/assets/datetime.py

```diff
@@ -0,0 +1,123 @@
+    if type(date) == np.datetime64:
+        date = date.astype("int64")
+    return date
+
+
+def get_dt_unit_factor(date, _type):
+    if type(date) == datetime.datetime:
+        date = date.timestamp()
+
+    _pow = dt[str(_type)] - int(math.log10(date_to_int(date)))
+    return math.pow(10, _pow)
+
+
+def to_datetime(dates):
+    """
+    Description:
+        Convert dates to pd.to_datetime while also figuring out
+        date_type unit
+    -------------------------------------------
+    Input:
+        dates = list or tuple of dates
+    -------------------------------------------
+    Output:
+        pd.to_datetime object
+    """
+    unit = {}
+    test_date = dates[0] if type(dates) in [list, tuple, pd.Series] else dates
+    if type(test_date) != datetime.datetime:
+        unit["unit"] = dt_unit[int(math.log10(date_to_int(test_date)))]
+    return pd.to_datetime(dates, **unit)
+
+
+def to_dt_if_datetime(dates, _type):
+    """
+    Description:
+        convert to datetime.datetime if _type in CUDF_DATETIME_TYPES
+
+    -------------------------------------------
+    Input:
+        dates = list or tuple of integer timestamps objects
+        _type = dtype
+    -------------------------------------------
+
+    Ouput:
+        list of datetime.datetime objects
+    """
+    if _type in CUDF_DATETIME_TYPES and type(dates) in [list, tuple]:
+        return type(dates)(to_datetime(dates).to_pydatetime())
+    return dates
+
+
+def to_np_dt64_if_datetime(dates, _type):
+    """
+    Description:
+        convert to np.datetime64 if _type in CUDF_DATETIME_TYPES
+
+    -------------------------------------------
+    Input:
+        dates = list or tuple of datetime.datetime objects
+        _type = dtype
+    -------------------------------------------
+
+    Ouput:
+        list of np.datetime64 objects
+
+    """
+    if _type in CUDF_DATETIME_TYPES and type(dates) in [list, tuple]:
+        dates = to_datetime(dates)
+        return type(dates)([date.to_datetime64() for date in dates])
+    return dates
+
+
+def check_series_for_nan(ser_):
+    if ser_.dtype == "float64":
+        return ser_[~cp.isnan(ser_)].shape[0] > 0
+    return True
+
+
+def to_int64_if_datetime(dates, _type):
+    """
+    Description:
+        convert to int64 for input to datashader based plots
+    -------------------------------------------
+    Input:
+        col: cudf.Series | list | tuple
+        _type: cudf.dtype
+    """
+    if _type in CUDF_DATETIME_TYPES:
+        if type(dates) in [list, tuple]:
+            # compute date seconds factor
+            dt_s_factor = get_dt_unit_factor(dates[0], _type)
+            return (np.array(dates).astype("int64")) * dt_s_factor
+        elif type(dates) == cudf.Series and check_series_for_nan(dates):
+            # compute date seconds factor
+            dt_s_factor = get_dt_unit_factor(dates.iloc[0], _type)
+            return (dates.astype("int64")) * dt_s_factor
+    return dates
```

Comment:
This comment does not have any specific actionable suggestion. I wish I had more fully formed thoughts, but I wanted to at least just jot down a consideration for the future. One thing that troubles me a bit is how dtypes are stored separately and redundantly. Things are either arrays (carry their own dtype) or not (no dtype) and storing this again only affords a chance for things to become out of sync. I guess instead of things roughly like
```
to_int64_if_datetime(event.geometry["x"], self.x_dtype)
```
I'd be happier to see more like
```
self.to_xaxis_type(event.geometry["x"])
```
where `to_xaxis_type` can inspect `self.source[self.x].dtype`  instead of requiring the dtype to actually be stored so it can be passed to some outside function (if a *convenience* is useful, `self.x_dtype` could be a *computed property*, instead of actually storing it as it does now). 

I know this oversimplifies things e.g dtypes are stored on charts and only reference the df on some *other* objects 
```
self.x_dtype = dashboard_cls._cuxfilter_df.data[self.x].dtype
```
but to me this only increases the likelihood that some future code changes could result in things becoming out of sync. Maybe a computed read-only property for `self.x_dtype` would be a potential quick improvement.

Reply:
converted self.x_dtype and self.y_dtype to read-only compyted properties  as seen [here](https://github.com/AjayThorve/cuxfilter/blob/new_dtypes/python/cuxfilter/charts/core/core_chart.py#L76-L90).

Also, transfered execution of all dt functions to BaseChart, so that something like `self.to_xaxis_type(event.geometry["x"])` works, all functions [here](https://github.com/AjayThorve/cuxfilter/blob/new_dtypes/python/cuxfilter/charts/core/core_chart.py#L132-L204)